### PR TITLE
Messages in drawer

### PR DESCRIPTION
**Pull request by eoji**
# What ❓
- Adding `Messages` to the navigation drawer in Discovery.
- When a user has an `unreadMessagesCount` > 0, we display the count to the right of the label.
- Since there are now 3 ways to get to the `MessageThreadsActivity`, we added `KoalaContext.Mailbox`. Users can access their mailbox from `creator_dashboard`, `drawer`, or `profile.
- `LoggedInViewHolder` didn't have a ViewModel so I added one so I could test!

# How to QA? 🤔
1. Open the app and verify the `Messages` row is only there for logged in users.
2. The unread count should display if it's greater than 0.
3. Clicking the row should take you to your inbox.

# Story 📖
[Trello Story](https://trello.com/c/toqzvUWn/1376-add-messages-to-android-drawer)

# See 👀

### CODE REVIEW

**Comment by eoji** [comment]
[Add Messages to android drawer](https://trello.com/c/toqzvUWn/1376-add-messages-to-android-drawer)

**Comment by Scollaco** [comment]
This is looking awesome! (And I could see I have 3 unread messages 😅 )

I noticed, though, that when I go to `Sent` messages,  I still see the `2 unread` text. Is it intended to be like that? Maybe we could show it only for messages in the `Inbox` , since we don't have unread sent messages.

Screenshot 2019-05-24 at 11 47 48

**Comment by eoji** [comment]
Yeah, I wasn't sure because we only get `unreadMessagesCount` and those unreads *could* be in your sent inbox also. Maybe we could parking lot during standup?

**Comment by eoji** [comment]
Sent mailbox doesn't show unread count

File: app/src/main/java/com/kickstarter/ui/viewholders/discoverydrawer/LoggedInViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.kickstarter.libs.utils.IntegerUtils
+import com.kickstarter.libs.utils.NumberUtils
+import com.kickstarter.libs.utils.ViewUtils
+import com.kickstarter.models.User
+import com.kickstarter.ui.viewholders.KSViewHolder
+import com.kickstarter.viewmodels.LoggedInViewHolderViewModel
+import com.squareup.picasso.Picasso
+import kotlinx.android.synthetic.main.discovery_drawer_logged_in_view.view.*
+
+class LoggedInViewHolder(@NonNull view: View, @NonNull private val delegate: Delegate) : KSViewHolder(view) {
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/ui/viewholders/discoverydrawer/LoggedInViewHolder.kt]: Shouldn't this class be `final`?

File: app/src/main/java/com/kickstarter/viewmodels/DiscoveryViewModel.java [hunk 2]
```diff
@@ -304,6 +310,7 @@ public ViewModel(final @NonNull Environment environment) {
     private final PublishSubject<Boolean> openDrawer = PublishSubject.create();
     private final PublishSubject<Integer> pagerSetPrimaryPage = PublishSubject.create();
     private final PublishSubject<NavigationDrawerData.Section.Row> parentFilterRowClick = PublishSubject.create();
+    private final PublishSubject<Void> messagesClick = PublishSubject.create();
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/viewmodels/DiscoveryViewModel.java]: Alphabetization here

File: app/src/main/java/com/kickstarter/viewmodels/DiscoveryViewModel.java [hunk 3]
```diff
@@ -347,6 +355,9 @@ public ViewModel(final @NonNull Environment environment) {
     @Override public void loggedInViewHolderProfileClick(final @NonNull LoggedInViewHolder viewHolder, final @NonNull User user) {
       this.profileClick.onNext(null);
     }
+    @Override public void loggedInViewHolderMessagesClick(final @NonNull LoggedInViewHolder viewHolder) {
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/viewmodels/DiscoveryViewModel.java]: And here

File: app/src/main/java/com/kickstarter/viewmodels/LoggedInViewHolderViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    interface Inputs {
+        /** Call with the current user when the data is bound to the view. */
+        fun configureWith(user: User)
+    }
+
+    interface Outputs {
+        /** Emits the user's medium avatar URL. */
+        fun avatarUrl(): Observable<String>
+
+        /** Emits a boolean determining if the user is a creator. */
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/viewmodels/LoggedInViewHolderViewModel.kt]: Here I would say that the Observable emits a boolean determining if the dashboardRow should be gone, depending if the user is creator or not.

File: app/src/main/java/com/kickstarter/viewmodels/LoggedInViewHolderViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        /** Emits the user to pass to delegate. */
+        fun user(): Observable<User>
+    }
+
+    class ViewModel(val environment: Environment) : ActivityViewModel<LoggedInViewHolder>(environment), Inputs, Outputs {
+
+        private val user = PublishSubject.create<User>()
+
+        private val avatarUrl = BehaviorSubject.create<String>()
+        private val name = BehaviorSubject.create<String>()
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/viewmodels/LoggedInViewHolderViewModel.kt]: Alphabetization here too.

File: app/src/main/java/com/kickstarter/viewmodels/MessageThreadsViewModel.java [hunk 6]
```diff
@@ -76,6 +78,9 @@
 
     /** Emits the unread message count to be displayed. */
     Observable<Integer> unreadMessagesCount();
+
+    /** Emits the unread message count to be displayed. */
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/viewmodels/MessageThreadsViewModel.java]: Copy pasta

File: app/src/main/java/com/kickstarter/ui/viewholders/discoverydrawer/LoggedInViewHolder.kt (hunk 1 above)
Comment (eoji) [reply app/src/main/java/com/kickstarter/ui/viewholders/discoverydrawer/LoggedInViewHolder.kt]: Nope!
Screen Shot 2019-05-28 at 10 17 31 AM

**Pull request merged**